crates/sound/src/audio_backends/alsa.rs

```diff
@@ -0,0 +1,550 @@
+/// Alsa backend
+//
+// Manos Pitsidianakis <manos.pitsidianakis@linaro.org>
+// SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
+use std::{
+    convert::{TryFrom, TryInto},
+    sync::mpsc::{channel, Receiver, Sender},
+    sync::{Arc, Mutex, RwLock},
+    thread,
+    thread::sleep,
+    time::Duration,
+};
+
+use alsa::{
+    pcm::{Access, Format, HwParams, State, PCM},
+    Direction, PollDescriptors, ValueOr,
+};
+use virtio_queue::Descriptor;
+use vm_memory::Bytes;
+
+use super::AudioBackend;
+use crate::{
+    device::ControlMessage,
+    stream::{PCMState, Stream},
+    virtio_sound::{
+        self, VirtioSndPcmSetParams, VIRTIO_SND_D_INPUT, VIRTIO_SND_D_OUTPUT, VIRTIO_SND_S_BAD_MSG,
+        VIRTIO_SND_S_NOT_SUPP,
+    },
+    Result as CrateResult,
+};
+
+type AResult<T> = std::result::Result<T, alsa::Error>;
+
+#[derive(Clone, Debug)]
+pub struct AlsaBackend {
+    sender: Arc<Mutex<Sender<AlsaAction>>>,
+}
+
+#[derive(Debug)]
+enum AlsaAction {
+    SetParameters(usize, ControlMessage),
+    Prepare(usize),
+    Release(usize, ControlMessage),
+    Start(usize),
+    Stop(usize),
+    Write(usize),
+    Read(usize),
+}
+
+fn update_pcm(
+    pcm_: &Arc<Mutex<PCM>>,
+    stream_id: usize,
+    streams: &RwLock<Vec<Stream>>,
+) -> AResult<()> {
+    *pcm_.lock().unwrap() = {
+        let streams = streams.read().unwrap();
+        let s = &streams[stream_id];
+        let pcm = PCM::new(
+            "default",
+            match s.direction {
+                d if d == VIRTIO_SND_D_OUTPUT => Direction::Playback,
+                d if d == VIRTIO_SND_D_INPUT => Direction::Capture,
+                other => panic!("Invalid virtio-sound stream: {}", other),
+            },
+            false,
+        )?;
+
+        {
+            let hwp = HwParams::any(&pcm)?;
+            hwp.set_channels(s.params.channels.into())?;
+            hwp.set_rate(
+                match s.params.rate {
```

Comment:
As this is something done also in the pipewire backend, i wonder if it really should be some enum or a wrapper around u8 with these constants and have a function that gives you the actual rate value.

Reply:
Agreed. I think it's early/unnecessary to do it in this PR, but I might do it next time I force push